plugins/com.google.cloud.tools.eclipse.dataflow.ui/src/com/google/cloud/tools/eclipse/dataflow/ui/preferences/RunOptionsDefaultsComponent.java

```diff
@@ -153,35 +153,136 @@ public RunOptionsDefaultsComponent(Composite target, int columns, MessageTarget
 
     // Staging Location, Combo, and Label occupy a single line
     comboLabel.setLayoutData(new GridData(SWT.END, SWT.CENTER, false, false, 1, 1));
-    stagingLocationInput.setLayoutData(new GridData(
-        SWT.FILL, SWT.CENTER, true, false, columns - STAGING_LOCATION_SPENT_COLUMNS, 1));
+    stagingLocationInput.setLayoutData(new GridData(SWT.FILL, SWT.CENTER, true, false,
+        columns - STAGING_LOCATION_SPENT_COLUMNS, 1));
 
     accountSelector.addSelectionListener(new Runnable() {
       @Override
       public void run() {
         // Don't use "removeAll()", as it will clear the text field too.
         stagingLocationInput.remove(0, stagingLocationInput.getItemCount() - 1);
         completionListener.setContents(ImmutableSortedSet.<String>of());
-        updateStagingLocations(getProject());
+        updateStagingLocations(getProject(), 0); // no delay
+        validate();
       }
     });
 
+    projectInput.addModifyListener(new ModifyListener() {
+      @Override
+      public void modifyText(ModifyEvent e) {
+        updateStagingLocations(getProject(), NEXT_KEY_DELAY_MS);
+        validate();
+      }
+    });
     projectInput.addFocusListener(new GetProjectStagingLocationsListener());
 
     completionListener = new SelectFirstMatchingPrefixListener(stagingLocationInput);
     stagingLocationInput.addModifyListener(completionListener);
     stagingLocationInput.addModifyListener(new ModifyListener() {
       @Override
       public void modifyText(ModifyEvent event) {
-        verifyStagingLocation(stagingLocationInput.getText());
+        startStagingLocationCheck(NEXT_KEY_DELAY_MS);
       }
     });
     createButton.addSelectionListener(new CreateStagingLocationListener());
 
-    stagingLocationInput.addModifyListener(new EnableCreateButton());
+    stagingLocationInput.addModifyListener(new ModifyListener() {
+      @Override
+      public void modifyText(ModifyEvent e) {
+        validate();
```

Comment:
nit: what if we move this `validate()` into the `addModifyListener()` code a little bit above?